File: code-library/12-panel-data/panel-data.py
# ==============================================================================
#  MODULE 12: PANEL DATA & REPEATED MEASURES
#  Dataset: CES 2020 (simulated two-wave panel)
#
#  Stack: linearmodels (FE/RE), statsmodels (DiD), pandas, matplotlib
#  pip install linearmodels
# ==============================================================================

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
from scipy import stats

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: CREATE SIMULATED TWO-WAVE PANEL
# ==============================================================================

wave1 = ces[[
    "caseid", "age", "sex", "education", "race_eth", "college", "white_nh",
    "census_region", "state_fips", "ideology5", "party_id3", "dem", "rep",
    "imm_restrict", "econ_retro", "voted", "wt_post"
]].copy()
wave1["wave"] = 0

rng   = np.random.default_rng(20240101)
wave2 = wave1.copy()
wave2["wave"] = 1

noise = rng.normal(0, 0.5, len(wave2))

# Simulate treatment effect: college-educated shift slightly more liberal
wave2["ideology5"] = np.where(
    wave2["college"] == 1,
    wave2["ideology5"] - 0.3 + noise,
    wave2["ideology5"] + noise
)
wave2["ideology5"] = wave2["ideology5"].clip(1, 5)

panel = pd.concat([wave1, wave2], ignore_index=True).sort_values(["caseid", "wave"])

print(f"Panel: {len(panel):,} observations, {panel['caseid'].nunique():,} units")


# ==============================================================================
# SECTION 2: POOLED OLS (BASELINE)
# ==============================================================================

from statsmodels.stats.sandwich_covariance import cov_cluster

m_pooled = smf.ols(
    "ideology5 ~ college + wave + age + C(sex)",
    data=panel
).fit()

# Cluster SEs by person
cluster_groups = panel.loc[m_pooled.model.data.orig_endog.index, "caseid"]
cov_cl  = cov_cluster(m_pooled, cluster_groups)
se_cl   = np.sqrt(np.diag(cov_cl))

print("Pooled OLS (clustered SEs):")
for param, coef, se in zip(m_pooled.params.index, m_pooled.params, se_cl):
    t = coef / se
    p = 2 * stats.t.sf(abs(t), df=m_pooled.df_resid)
    print(f"  {param:<30} {coef:>8.4f}  ({se:.4f})  p={p:.4f}")


# ==============================================================================
# SECTION 3: FIXED EFFECTS (WITHIN ESTIMATOR) WITH LINEARMODELS
# ==============================================================================

# linearmodels.panel.PanelOLS provides FE/RE with correct panel SEs.
# Must set a MultiIndex (entity, time) before fitting.

try:
    from linearmodels.panel import PanelOLS, RandomEffects, BetweenOLS
    from linearmodels import PanelData

    panel_indexed = panel.set_index(["caseid", "wave"])

    # Fixed effects: absorb individual-level FE
    # entity_effects=True = demean within person (equivalent to Stata's xtreg, fe)
    m_fe = PanelOLS(
        dependent     = panel_indexed["ideology5"],
        exog          = panel_indexed[["college", "wave"]],
        entity_effects = True,
        time_effects   = False
    ).fit(cov_type="clustered", cluster_entity=True)

    print("\nFixed Effects Model:")
    print(m_fe.summary.tables[1])   # coefficient table

    # Note: age and sex are time-invariant → absorbed by entity FE → dropped
    # Only time-varying predictors remain

    # --- Random Effects ---
    m_re = RandomEffects(
        dependent = panel_indexed["ideology5"],
        exog      = panel_indexed[["college", "wave", "age"]]
    ).fit(cov_type="robust")

    print("\nRandom Effects Model:")
    print(m_re.summary.tables[1])

    # --- Hausman-style comparison (linearmodels) ---
    fe_params = m_fe.params
    re_params = m_re.params

    common = [p for p in fe_params.index if p in re_params.index]
    diff   = fe_params[common] - re_params[common]
    print(f"\nHausman-style diff (FE - RE) for college:")
    print(f"  {diff.get('college', 'N/A'):.4f}")
    print("  Large difference → prefer FE for causal inference")

except ImportError:
    print("linearmodels not installed — run: pip install linearmodels")
    m_fe = None


# ==============================================================================
# SECTION 4: FIRST DIFFERENCES (MANUAL)
# ==============================================================================

panel_fd = (
    panel.sort_values(["caseid", "wave"])
    .groupby("caseid")
    .apply(lambda g: g.set_index("wave").diff().loc[1], include_groups=False)
    .dropna(subset=["ideology5", "college"])
    .reset_index()
)

m_fd = smf.ols(
    "ideology5 ~ college",
    data=panel_fd
).fit()

# Cluster SEs by person
cluster_fd = panel_fd.loc[m_fd.model.data.orig_endog.index, "caseid"]
cov_fd     = cov_cluster(m_fd, cluster_fd)
se_fd      = np.sqrt(np.diag(cov_fd))

print("\nFirst Differences:")
for param, coef, se in zip(m_fd.params.index, m_fd.params, se_fd):
    t = coef / se
    p = 2 * stats.t.sf(abs(t), df=m_fd.df_resid)
    print(f"  {param:<20} {coef:>8.4f}  ({se:.4f})  p={p:.4f}")


# ==============================================================================
# SECTION 5: DIFFERENCE-IN-DIFFERENCES (DiD)
# ==============================================================================

panel_did = panel.assign(
    treated = panel["college"],
    post    = panel["wave"],
    did     = panel["college"] * panel["wave"]
)

# Standard DiD regression
m_did = smf.ols(
    "ideology5 ~ treated + post + did",
    data=panel_did
).fit()

cluster_did = panel_did.loc[m_did.model.data.orig_endog.index, "caseid"]
cov_did     = cov_cluster(m_did, cluster_did)
se_did      = np.sqrt(np.diag(cov_did))

print("\nDiD Regression:")
for param, coef, se in zip(m_did.params.index, m_did.params, se_did):
    t = coef / se
    p = 2 * stats.t.sf(abs(t), df=m_did.df_resid)
    stars = "***" if p < .001 else "**" if p < .01 else "*" if p < .05 else ""
    print(f"  {param:<20} {coef:>8.4f}{stars}  ({se:.4f})  p={p:.4f}")

print("\n  Interpretation: 'did' coefficient is the ATT (Average Treatment")
print("  Effect on the Treated) under the parallel trends assumption.")

# DiD with controls
m_did_ctrl = smf.ols(
    "ideology5 ~ treated * post + age + C(sex) + C(census_region)",
    data=panel_did
).fit()

cluster_ctrl = panel_did.loc[m_did_ctrl.model.data.orig_endog.index, "caseid"]
cov_ctrl     = cov_cluster(m_did_ctrl, cluster_ctrl)
se_ctrl      = np.sqrt(np.diag(cov_ctrl))

print("\nDiD with Controls — 'treated:post' (ATT):")
param  = "treated:post"
if param in m_did_ctrl.params:
    coef = m_did_ctrl.params[param]
    idx  = list(m_did_ctrl.params.index).index(param)
    se   = se_ctrl[idx]
    t    = coef / se
    p    = 2 * stats.t.sf(abs(t), df=m_did_ctrl.df_resid)
    print(f"  {coef:.4f} (SE={se:.4f}, p={p:.4f})")


# ==============================================================================
# SECTION 6: PARALLEL TRENDS PLOT
# ==============================================================================

trend = (
    panel_did.groupby(["treated", "wave"])["ideology5"]
    .mean()
    .reset_index()
    .rename(columns={"ideology5": "mean_ideo"})
)

fig, ax = plt.subplots(figsize=(7, 5))
colors = {0: "navy", 1: "firebrick"}
labels = {0: "No College (Control)", 1: "College (Treated)"}

for treated_val, grp in trend.groupby("treated"):
    ax.plot(grp["wave"], grp["mean_ideo"],
            color=colors[treated_val], label=labels[treated_val],
            linewidth=2, marker="o", markersize=8)

ax.axvline(0.5, color="gray", linestyle="--", linewidth=1, label="Treatment cutpoint")
ax.set_xticks([0, 1])
ax.set_xticklabels(["Pre", "Post"])
ax.set_xlabel("")
ax.set_ylabel("Mean Ideology (1=Very Lib, 5=Very Con)")
ax.set_title("Parallel Trends Check\nMean Ideology by Group and Wave")
ax.legend()
plt.tight_layout()
plt.savefig("parallel_trends.png", dpi=300)
plt.close()


# ==============================================================================
# SECTION 7: WITHIN VS. BETWEEN VARIANCE
# ==============================================================================

# Decompose variance into within-person and between-person components
person_means = panel.groupby("caseid")["ideology5"].mean()
panel = panel.merge(person_means.rename("ideo_person_mean"), on="caseid")

panel["ideo_within"]  = panel["ideology5"] - panel["ideo_person_mean"]
panel["ideo_between"] = panel["ideo_person_mean"] - panel["ideology5"].mean()

var_total   = panel["ideology5"].var()
var_within  = panel["ideo_within"].var()
var_between = panel["ideo_between"].var()

print(f"\nVariance decomposition — ideology5:")
print(f"  Total:   {var_total:.4f}")
print(f"  Between: {var_between:.4f} ({var_between/var_total*100:.1f}%)")
print(f"  Within:  {var_within:.4f}  ({var_within/var_total*100:.1f}%)")
print("  Fixed effects uses only the within-person variance.")

panel = panel.drop(columns=["ideo_person_mean", "ideo_within", "ideo_between"])


# ==============================================================================
# SECTION 8: SUMMARY TABLE
# ==============================================================================

print("\n" + "="*60)
print("MODEL SUMMARY")
print("="*60)
print(f"{'Model':<22} {'College coef':>14} {'N':>8}")
print("-"*60)

m_pool_cluster_coef = m_pooled.params.get("college", np.nan)
print(f"{'Pooled OLS':<22} {m_pool_cluster_coef:>14.4f} {int(m_pooled.nobs):>8,}")

if m_fe is not None and "college" in m_fe.params:
    print(f"{'Fixed Effects':<22} {m_fe.params['college']:>14.4f} {int(m_fe.nobs):>8,}")

fd_coef = m_fd.params.get("college", np.nan)
print(f"{'First Differences':<22} {fd_coef:>14.4f} {int(m_fd.nobs):>8,}")

did_coef = m_did.params.get("did", np.nan)
print(f"{'DiD (treatment×post)':<22} {did_coef:>14.4f} {int(m_did.nobs):>8,}")

print("\nModule 12 complete.")

File: code-library/12-panel-data/panel-data.R
# ==============================================================================
#  MODULE 12: PANEL DATA & REPEATED MEASURES
#  Dataset: CES 2020 (simulated two-wave panel)
#
#  Purpose: Fixed effects, random effects, and difference-in-differences
#  using plm (panel linear models) and lfe (large fixed effects).
#
#  install.packages(c("plm", "lfe", "fixest"))
#  fixest is the modern, fastest option for FE models — use it for large data.
# ==============================================================================

library(tidyverse)
library(plm)         # panel linear models (classic approach)
library(fixest)      # fast fixed effects (modern approach — recommended)
library(broom)
library(modelsummary)

ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: CREATE A SIMULATED TWO-WAVE PANEL
# ==============================================================================

# Wave 1: baseline (pre-election)
wave1 <- ces |>
  select(caseid, age, sex, education, race_eth, college, white_nh,
         census_region, state_fips, ideology5, party_id3, dem, rep,
         imm_restrict, econ_retro, voted, wt_post) |>
  mutate(wave = 0)

# Wave 2: post-election — simulate change in ideology for college respondents
set.seed(20240101)

wave2 <- wave1 |>
  mutate(
    wave      = 1,
    ideo_noise = rnorm(n(), 0, 0.5),
    # Simulate treatment effect: college-educated shift slightly more liberal
    ideology5 = case_when(
      college == 1 ~ ideology5 - 0.3 + ideo_noise,
      TRUE         ~ ideology5 + ideo_noise
    ),
    ideology5 = pmax(1, pmin(5, ideology5))   # clamp to [1, 5]
  ) |>
  select(-ideo_noise)

panel_df <- bind_rows(wave1, wave2) |>
  arrange(caseid, wave)

cat("Panel observations:", nrow(panel_df), "\n")
cat("Unique respondents:", n_distinct(panel_df$caseid), "\n")


# ==============================================================================
# SECTION 2: DECLARE PANEL STRUCTURE WITH PLM
# ==============================================================================

# pdata.frame() declares the panel structure — sets index (unit, time)
panel_plm <- pdata.frame(panel_df, index = c("caseid", "wave"))

# pdim() summarizes the panel structure
pdim(panel_plm)

# Within vs. between variance for key variables
summary(Between(panel_plm$ideology5))   # between-person variance
summary(Within(panel_plm$ideology5))    # within-person variance (over waves)


# ==============================================================================
# SECTION 3: POOLED OLS (BASELINE)
# ==============================================================================

# Ignores panel structure — standard lm() with clustered SEs on person ID
m_pooled <- lm(ideology5 ~ college + wave + age + factor(sex),
               data = panel_df)

# Cluster SEs by person (caseid) using sandwich
library(sandwich)
library(lmtest)
coeftest(m_pooled, vcov = vcovCL(m_pooled, cluster = ~ caseid))


# ==============================================================================
# SECTION 4: FIXED EFFECTS MODEL
# ==============================================================================

# plm with model = "within" estimates the within (FE) estimator
# FE removes all time-invariant variation: age, sex, race drop out automatically

m_fe_plm <- plm(
  ideology5 ~ college + wave,
  data  = panel_plm,
  model = "within",
  effect = "individual"   # individual FE (demean within person)
)

summary(m_fe_plm)

# Clustered SEs (cluster on individual)
coeftest(m_fe_plm, vcov = plm::vcovHC(m_fe_plm, type = "HC1", cluster = "group"))

# --- Using fixest (faster, modern alternative) ---
m_fe_fixest <- feols(
  ideology5 ~ college + wave | caseid,   # | caseid = absorb individual FE
  data    = panel_df,
  cluster = ~ caseid
)

summary(m_fe_fixest)
# fixest absorbs fixed effects efficiently — critical for large panels


# ==============================================================================
# SECTION 5: RANDOM EFFECTS MODEL
# ==============================================================================

# plm with model = "random" estimates the GLS random effects estimator
# RE assumes unit effects are UNCORRELATED with predictors (strong assumption)
# Advantage over FE: can include time-invariant predictors (sex, race, etc.)

m_re <- plm(
  ideology5 ~ college + wave + age + factor(sex),
  data  = panel_plm,
  model = "random"
)

summary(m_re)


# ==============================================================================
# SECTION 6: HAUSMAN TEST — FE VS. RE
# ==============================================================================

# phtest() runs the Hausman test comparing FE and RE coefficients
# H0: RE is consistent (FE and RE give similar estimates)
# Reject H0 → use FE

phtest(m_fe_plm, m_re)

# Significant p-value: RE is inconsistent (endogeneity in random effects)
# → prefer fixed effects for causal inference


# ==============================================================================
# SECTION 7: FIRST DIFFERENCES
# ==============================================================================

# plm with model = "fd" computes first-difference estimator
# With T=2, FD and FE give identical estimates

m_fd <- plm(
  ideology5 ~ college + wave,
  data  = panel_plm,
  model = "fd"
)

summary(m_fd)

# Manual first differences using dplyr
panel_fd <- panel_df |>
  arrange(caseid, wave) |>
  group_by(caseid) |>
  mutate(
    d_ideology5 = ideology5 - lag(ideology5),
    d_college   = college   - lag(college)
  ) |>
  filter(wave == 1) |>   # keep only the difference rows
  ungroup()

lm(d_ideology5 ~ d_college, data = panel_fd) |>
  coeftest(vcov = vcovHC(lm(d_ideology5 ~ d_college, data = panel_fd),
                         type = "HC3"))


# ==============================================================================
# SECTION 8: DIFFERENCE-IN-DIFFERENCES (DiD)
# ==============================================================================

# DiD setup:
#   Treated = college == 1
#   Post    = wave == 1
#   DiD     = Treated × Post interaction
#
# The β on the interaction is the Average Treatment Effect on the Treated (ATT),
# under the parallel trends assumption.

panel_did <- panel_df |>
  mutate(
    treated = college,
    post    = wave,
    did     = treated * post
  )

# Standard DiD regression
m_did <- lm(
  ideology5 ~ treated + post + did,
  data = panel_did
)

coeftest(m_did, vcov = vcovCL(m_did, cluster = ~ caseid))

# Equivalent using * notation (includes main effects + interaction)
m_did2 <- lm(
  ideology5 ~ treated * post,
  data = panel_did
)

coeftest(m_did2, vcov = vcovCL(m_did2, cluster = ~ caseid))

# DiD with additional controls
m_did_controls <- lm(
  ideology5 ~ treated * post + age + factor(sex) + factor(census_region),
  data = panel_did
)

coeftest(m_did_controls, vcov = vcovCL(m_did_controls, cluster = ~ caseid))

# fixest version (cleanest syntax, handles clustering automatically)
m_did_fixest <- feols(
  ideology5 ~ i(post, treated, ref = 0) +    # interaction with reference period
    age + factor(sex),
  data    = panel_did,
  cluster = ~ caseid
)

summary(m_did_fixest)


# ==============================================================================
# SECTION 9: PARALLEL TRENDS PLOT
# ==============================================================================

trend_df <- panel_did |>
  group_by(treated, wave) |>
  summarise(mean_ideo = mean(ideology5, na.rm = TRUE), .groups = "drop")

ggplot(trend_df, aes(x = wave, y = mean_ideo,
                     color = factor(treated),
                     group = factor(treated))) +
  geom_line(linewidth = 1) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.5, linetype = "dashed", color = "gray50") +
  scale_x_continuous(breaks = c(0, 1), labels = c("Pre", "Post")) +
  scale_color_manual(values = c("navy", "firebrick"),
                     labels = c("No College (Control)", "College (Treated)"),
                     name   = NULL) +
  labs(
    title    = "Parallel Trends Check",
    subtitle = "Mean Ideology by Treatment Group and Wave",
    x        = NULL,
    y        = "Mean Ideology (1=Very Liberal, 5=Very Conservative)"
  ) +
  theme_minimal() +
  theme(legend.position = "bottom")

ggsave("parallel_trends.png", width = 7, height = 5, dpi = 300)


# ==============================================================================
# SECTION 10: RESULTS TABLE
# ==============================================================================

# RE model (plm object) doesn't support sandwich cluster SEs via modelsummary;
# show it separately and include only lm-based models in the joint table.
tidy(m_re, conf.int = TRUE)

models_panel <- list(
  "Pooled OLS" = m_pooled,
  "DiD"        = m_did_controls
)

modelsummary(
  models_panel,
  vcov    = ~ caseid,    # cluster SEs on person for all models
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  title   = "Panel Models: Effect of College on Ideology"
)

# FE model separately (fixest object)
modelsummary(
  list("Fixed Effects" = m_fe_fixest),
  stars = c("*" = .05, "**" = .01, "***" = .001)
)

message("Module 12 complete.")
